integration-tests/deployment/keystone/update_nodes.go

```diff
@@ -0,0 +1,194 @@
+		}
+	}
+
+	for _, p2pID := range p2pIds {
+		// read the existing capabilities for the node
+		info, err := registry.GetNode(&bind.CallOpts{}, p2pID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get node info for %s: %w", p2pID, err)
+		}
+		mergedCaps := make([]kcr.CapabilitiesRegistryCapability, 0)
+		// we only have the id; need to fetch the capabilities details
+		for _, capID := range info.HashedCapabilityIds {
+			cap, exists := capMap[capID]
+			if !exists {
+				return nil, fmt.Errorf("capability not found for %s", capID)
+			}
+			mergedCaps = append(mergedCaps, cap)
+		}
+		// append the new capabilities and dedup
+		mergedCaps = append(mergedCaps, capabilities...)
+		var deduped []kcr.CapabilitiesRegistryCapability
+		seen := make(map[string]struct{})
+		for _, cap := range mergedCaps {
+			if _, ok := seen[CapabilityID(cap)]; !ok {
+				seen[CapabilityID(cap)] = struct{}{}
+				deduped = append(deduped, cap)
+			}
+		}
+		out[p2pID] = deduped
+	}
+	return out, nil
+}
+
+func makeNodeParams(registry *kcr.CapabilitiesRegistry,
+	nopToNodes map[kcr.CapabilitiesRegistryNodeOperator][]*P2PSigner,
+	p2pToCapabilities map[p2pkey.PeerID][]kcr.CapabilitiesRegistryCapability) ([]kcr.CapabilitiesRegistryNodeParams, error) {
+
+	out := make([]kcr.CapabilitiesRegistryNodeParams, 0)
+	// get all the node operators from chain
+	registeredNops, err := registry.GetNodeOperators(&bind.CallOpts{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get node operators: %w", err)
+	}
+
+	// make a cache of capability from chain
+	var allCaps []kcr.CapabilitiesRegistryCapability
+	for _, caps := range p2pToCapabilities {
+		allCaps = append(allCaps, caps...)
+	}
+	capMap, err := fetchCapabilityIDs(registry, allCaps)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch capability ids: %w", err)
+	}
+
+	// flatten the onchain state to list of node params filtered by the input nops and nodes
+	for idx, rnop := range registeredNops {
+		// nop id is 1-indexed. no way to get value from chain. must infer from index
+		nopID := uint32(idx + 1)
+		nodes, ok := nopToNodes[rnop]
+		if !ok {
+			continue
+		}
+		for _, node := range nodes {
+			caps, ok := p2pToCapabilities[node.P2PKey]
+			if !ok {
+				return nil, fmt.Errorf("capabilities not found for node %s", node.P2PKey)
+			}
+			hashedCaps := make([][32]byte, len(caps))
+			for i, cap := range caps {
+				hashedCap, exists := capMap[CapabilityID(cap)]
+				if !exists {
+					return nil, fmt.Errorf("capability id not found for %s", CapabilityID(cap))
+				}
+				hashedCaps[i] = hashedCap
+			}
+			out = append(out, kcr.CapabilitiesRegistryNodeParams{
+				NodeOperatorId:      nopID,
+				P2pId:               node.P2PKey,
+				HashedCapabilityIds: hashedCaps,
+				Signer:              node.Signer,
+			})
+		}
+	}
+
+	return out, nil
+}
+
+func CapabilityID(c kcr.CapabilitiesRegistryCapability) string {
+	return fmt.Sprintf("%s_%s_%d", c.LabelledName, c.Version, c.CapabilityType)
```

Comment:
Is the intent for this to be a different type of CapabilityID from [this](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.8/keystone/CapabilitiesRegistry.sol#L170) ?

Reply:
thanks for pointing out. fixed and new comment